File: src/main/java/ru/arc/network/NetworkRegistry.java
package ru.arc.network;

import ru.arc.ai.assistant.ToolMessanger;
import ru.arc.hooks.HookRegistry;
import ru.arc.hooks.lands.LandsMessager;
import ru.arc.hooks.zauction.AuctionMessager;
import ru.arc.stock.HistoryManager;
import ru.arc.stock.HistoryMessager;
import ru.arc.xserver.playerlist.PlayerListMessager;

public class NetworkRegistry {

    private final RedisManager redisManager;

    public NetworkRegistry(RedisManager redisManager) {
        this.redisManager = redisManager;
    }
    public static LandsMessager landsMessager;
    PlayerListMessager playerListMessager;

    public void init() {
        playerListMessager = new PlayerListMessager("arc.proxy_player_list");
        redisManager.registerChannelUnique(playerListMessager.channel(), playerListMessager);

        landsMessager = new LandsMessager(redisManager, "arc.lands_req", "arc.lands_response");
        landsMessager.init();
        redisManager.registerChannelUnique(landsMessager.getRespChannel(), landsMessager);
        redisManager.registerChannelUnique(landsMessager.getReqChannel(), landsMessager);

        HistoryMessager historyMessager = new HistoryMessager("arc.high_lows_update", redisManager);
        redisManager.registerChannelUnique(historyMessager.channel, historyMessager);
        HistoryManager.setMessager(historyMessager);

        if (HookRegistry.auctionHook != null) {
            AuctionMessager auctionMessager = new AuctionMessager("arc.auction_items", "arc.auction_items_all", redisManager);
            redisManager.registerChannelUnique(auctionMessager.channel, auctionMessager);
            redisManager.registerChannelUnique(auctionMessager.channelAll, auctionMessager);
            HookRegistry.auctionHook.setAuctionMessager(auctionMessager);
        }

        ToolMessanger toolMessanger = new ToolMessanger();
        redisManager.registerChannelUnique(ToolMessanger.CHANNEL_REQUEST_TOOLS, toolMessanger);

        redisManager.init();
    }

}

File: src/main/kotlin/ru/arc/network/NetworkRegistry.kt
package ru.arc.network

import ru.arc.hooks.HookRegistry
import ru.arc.hooks.lands.LandsMessager
import ru.arc.hooks.zauction.AuctionMessager
import ru.arc.redis.ChannelListener
import ru.arc.redis.RedisOperations
import ru.arc.stock.HistoryManager
import ru.arc.stock.HistoryMessager
import ru.arc.xserver.playerlist.PlayerListMessager

class NetworkRegistry(
    private val redis: RedisOperations,
) : AutoCloseable {
    private val registrations = mutableListOf<Pair<String, ChannelListener>>()
    private var landBridge: LandsMessager? = null
    private var historyBridge: HistoryMessager? = null
    private var auctionBridge: AuctionMessager? = null

    companion object {
        @JvmField
        var landsMessager: LandsMessager? = null
    }

    @Synchronized
    fun init() {
        close()
        try {
            val playerList = PlayerListMessager("arc.proxy_player_list")
            register(playerList.channel, playerList)

            val lands = LandsMessager(redis, "arc.lands_req", "arc.lands_response")
            lands.init()
            landBridge = lands
            landsMessager = lands
            register(lands.respChannel, lands)
            register(lands.reqChannel, lands)

            val history = HistoryMessager("arc.high_lows_update", redis)
            historyBridge = history
            register(history.channel, history)
            HistoryManager.setMessager(history)

            HookRegistry.auctionHook?.let { hook ->
                val auction = AuctionMessager("arc.auction_items", "arc.auction_items_all", redis)
                auctionBridge = auction
                register(auction.channel, auction)
                register(auction.channelAll, auction)
                hook.auctionMessager = auction
            }
        } catch (error: Exception) {
            close()
            throw error
        }
    }

    @Synchronized
    override fun close() {
        registrations.asReversed().forEach { (channel, listener) ->
            redis.unregisterChannel(channel, listener)
        }
        registrations.clear()

        landBridge?.close()
        if (landsMessager === landBridge) landsMessager = null
        landBridge = null

        historyBridge?.let(HistoryManager::clearMessager)
        historyBridge = null

        auctionBridge?.let { bridge ->
            HookRegistry.auctionHook?.let { hook ->
                if (hook.auctionMessager === bridge) hook.auctionMessager = null
            }
        }
        auctionBridge = null
    }

    fun shutdown() = close()

    private fun register(channel: String, listener: ChannelListener) {
        redis.registerChannelUnique(channel, listener)
        registrations += channel to listener
    }
}
